src/main/java/org/jenkinsci/plugins/docker/commons/impl/CompositeKeyMaterialFactory.java

```diff
@@ -2,78 +2,109 @@
-                    throw new IOException("Error materializing credentials.", e);
-                }
-            } finally {
-                if (rhsMaterialized != null) {
-                    rhsMaterialized.close();
+        KeyMaterial[] keyMaterials = new KeyMaterial[factories.length];
+        EnvVars env = new EnvVars();
+        try {
+            for (int index = 0; index < factories.length; index++) {
+                keyMaterials[index] = factories[index].materialize();
+                env.putAll(keyMaterials[index].env());
+            }
+            return new CompositeKeyMaterial(env, keyMaterials);
+        } catch (Throwable e) {
+            for (int index = keyMaterials.length - 1; index >= 0; index--) {
+                try {
+                    if (keyMaterials[index] != null) {
+                        keyMaterials[index].close();
+                    }
+                } catch (IOException ioe) {
+                    // ignore
                 }
             }
+            // TODO Java 7+ use chained exceptions
+            if (e instanceof IOException) {
+                throw (IOException) e;
+            } else if (e instanceof InterruptedException) {
+                throw (InterruptedException) e;
+            } else if (e instanceof RuntimeException) {
+                throw (RuntimeException) e;
+            } else {
+                throw new IOException("Error materializing credentials.", e);
+            }
         }
     }
-    
+
     private static final class CompositeKeyMaterial extends KeyMaterial implements Serializable {
-        
+
         private static final long serialVersionUID = 1L;
 
-        private final KeyMaterial lhsMaterialized;
-        private final KeyMaterial rhsMaterialized;
+        private final KeyMaterial[] keyMaterials;
 
-        protected CompositeKeyMaterial(EnvVars envVars, KeyMaterial lhsMaterialized, KeyMaterial rhsMaterialized) {
+        protected CompositeKeyMaterial(EnvVars envVars, KeyMaterial... keyMaterials) {
             super(envVars);
-            this.lhsMaterialized = lhsMaterialized;
-            this.rhsMaterialized = rhsMaterialized;
+            this.keyMaterials = keyMaterials;
         }
 
         @Override
         public void close() throws IOException {
-            try {
-                lhsMaterialized.close();
-            } finally {
-                rhsMaterialized.close();
+            Throwable first = null;
+            for (int index = keyMaterials.length - 1; index >= 0; index--) {
+                try {
+                    if (keyMaterials[index] != null) {
+                        keyMaterials[index].close();
+                    }
+                } catch (IOException e) {
+                    first = first == null ? e : first;
+                } catch (RuntimeException e) {
+                    first = first == null ? e : first;
+                } catch (Throwable e) {
+                    first = first == null ? e : first;
```

Comment:
because I like to enumerate the kinds I expect rather than just a blind catch `Throwable`... so that later readers know I am intentionally catching `Throwable` rather than just being lazy